doc/hacking/style_guide.md

```diff
@@ -0,0 +1,187 @@
+    installed into the same path.
+- **Functions**: `lower_snake_case_without_prefix`
+- **Local variables**: `lower_snake_case_without_prefix`
+- **Member variables**: `_lower_snake_case_with_initial_score`
+
+Indentation
+-----------
+
+Always use two spaces to indent.  Tabs should never appear in a C++ file.
+
+Access modifiers (`public`/`private`/`protected`/`PUBLISHED`) are never
+indented, and do not affect the indentation of the surrounding lines.
+
+Spaces
+------
+
+Always put one space after a control keyword, before the condition's
+parentheses.  Don't pad the inside of parentheses with spaces unless there are
+several nested parenthetical expressions and the extra spaces contribute
+substantially to readability.
+
+    if (x == y) {
+
+Don't place a space after a function name (declarations, definitions, or calls).
+
+    void function_name();
+    ...
+    function_name();
+
+Always place a space after each comma in an arguments list.
+
+    function_call(x, y, z);
+
+Always pad binary operators with a space on either side.
+
+    a || b
+
+Never end a line with trailing whitespace.
+
+Function definitions
+--------------------
+
+The recipe for a function definition is:
+
+1. JavaDoc comment describing the purpose of the function
+2. Function return value and class
+3. **At the beginning of its own line,** the function's name, followed by the
+   parameter list and anything else necessary for the function's signature.
+4. Opening brace on the same line, not its own line.
+5. Single empty line before next definition.
+
+For example:
+
+    /**
+     * String documenting the function, in JavaDoc style.
+     */
+    void ClassName::
+    function_name(int x, int y) const {
+      ...
+    }
+
+    /**
+     * ...
+     */
+    ...
+
+If the function is a constructor with an initializer list, the `:` goes on the
+same line as the constructor name, and the initializers are listed *one per
+line*, indented by two spaces.
+
+    ClassName::
+    ClassName() :
+      _a("a"),
+      _b("b"),
+      _c("c")
+    {
+    }
+
+Empty function bodies in a .cxx/.I file still contain a single line break (i.e.
+to put `}` on its own line).  In a .h file, an empty function can be given as
+`{}`
+
+Braces
+------
+
+Always put an opening `{` on the same line as the statement that calls for the
+brace.  There is exactly one exception to this: constructors with initializer
+lists.
+
+    SomeClass::
+    SomeClass() :
+      _one(1),
+      _two(2)
+    {
+      if (_one < _two) {
+        ...
+      }
+    }
+
+Comments
+--------
+
+Comments should be properly-formatted English, with proper capitalization and
+punctuation rules.  There are two spaces between sentences.
+
+Note that `//` is preferred over `/*` except in some special circumstances.
+
+Line limits and continuations
+-----------------------------
+
+Lines should be limited to 80 columns.  This is not a hard limit, but if a line
+does go over this length, consider a continuation.
```

Comment:
We can, although if we're making a decision over something that small (78 vs. 80) we should be discussing the strictness of the line limit first.

clang-format allows setting soft constraints on the line limit, where you can tell it that it's OK to go a little over if the alternative would be adding line breaks in awkward spots.

We should also be figuring out whether it's necessary to rewrap something that's under the limit - documentation comments wrapped at 78 do not violate the 80-column limit unless we require using as few lines as possible.

Reply:
I don't think the line limit should be strict, especially as setting arbitrary limits can actually hurt readability.  An exception is flowing text such as documentation strings, which is easy to wrap and this is currently done at 78 columns.

Reply:
I see. So you're suggesting a soft limit for code and a hard limit for comments.

However I'm still not sure how raising the limit from 78 to 80 requires a rewrap. If a comment fits in 78 columns (as they all do currently), it necessarily also fits in 80 (as the current draft proposes). The difference only seems relevant if we want to set the additional rule of using all available space in the line before wrapping.

Reply:
I think the most reasonable would be to set the limit to 80 characters, and only ensure that a line is never longer than that. In comments you will rarely hit exactly 80 characters. For example: 

```
/**
 * Replaces the current system-RAM image with the new data, converting it       // 73
 * first if necessary from the indicated component-order format.  See           // 79
 * get_ram_image_as() for specifications about the format.  This method cannot  // 78
 * support compressed image data or sub-pages; use set_ram_image() for that.    // 76
 */
```

If we would add a rule that a comment line always has to be 80 characters, this would
require having to fill the remaining space with spaces or so .. that would be really weird (And would introduce trailing spaces).

This way, its fine to have docstrings with any wrap (Currently 78) and never more than 80 characters long per line.

Reply:
@cfsworks: changing the limit from 78 to 80 means that running a tool that is programmed to rewrap to 80 would make lots of unnecessary changes to documentation strings.  I just see not much reason to change the rule again since we have previously explicitly run a tool to rewrap everything to 78.

…but maybe this is just my OCD-ishness talking, so if you both feel we should make it 80 characters, I'm willing to be overruled on this.